Add a summary list example where only some rows have actions

36degrees opened the pull request:
**The guidance will not match the HTML in the example until updated to GOV.UK Frontend v4.0 with the changes from https://github.com/alphagov/govuk-frontend/pull/2323.** Do not merge until after the Design System has been updated to 4.0.

The example also takes into account changes to the other existing examples made in #1989, which has not yet been merged.

Closes #1948.

calvin-lau-sig7 commented on src/components/summary-list/index.md.njk:
```diff
@@ -30,24 +30,31 @@ Do not use it for tabular data or a simple list of information or tasks, like a
+### Removing the borders
 
-### Summary list without borders
+The summary list includes separating borders to help users read each row and its action.
```
I know adding this explanation is a bit clunky, but it looked like adding a "do not do this" line afterwards started to look a bit confusing (and easy to miss if it went after the example).

calvin-lau-sig7 commented on src/components/summary-list/index.md.njk:
```diff
@@ -30,24 +30,31 @@ Do not use it for tabular data or a simple list of information or tasks, like a
+If your summary list does not have any actions, you can choose to remove the separating borders with the `govuk-summary-list--no-border` class.
 
 {{ example({group: "components", item: "summary-list", example: "without-borders", html: true, nunjucks: true, open: false}) }}
 
```
Not a big deal, but I'm not sure I understand when someone would choose to selectively remove borders for rows... 
- To show if a row has an action or not? Seems like it'd look a bit off
- To group rows together? That'd only be a visual-only distinction which would not met WCAG

36degrees replied:
Probably the most common is to remove the bottom border from the very last row – we even have [an issue to make that the default behaviour](https://github.com/alphagov/govuk-frontend/issues/1909).

Not sure we know enough about other use cases to go into detail really. Unsure about whether grouping being visual only would be a WCAG fail… likely depends how important the grouping is, and whether it is communicated another way e.g. through the labels used.

For example, this sort of thing would probably be OK…

```
Home address           1234 Lower Avenue           Change
Home phone number      01234 567890                Change
---------------------------------------------------------
Work address           21 High Street              Change
Work phone number      01234 098765                Change
---------------------------------------------------------
```

calvin-lau-sig7 replied:
Ahh okay, thanks for explaining. They'd be grouped in that instance (as similar/repeated) field names.

calvin-lau-sig7 commented on src/components/summary-list/index.md.njk:
```diff
@@ -30,24 +30,31 @@ Do not use it for tabular data or a simple list of information or tasks, like a
 {{ example({group: "components", item: "summary-list", example: "default", html: true, nunjucks: true, open: false, titleSuffix: "second"}) }}
 
-### Summary list without actions
+### Adding actions to only some rows
```
Alternatives?
- Adding actions to some rows, but not others

EoinShaughnessy replied:
Is line 49 about adding actions to some rows, or is it more about indicating which rows do not have actions?

calvin-lau-sig7 replied:
It's... both!

The way Ollie explained it, if there is no action (link text) for a particularly row, there also needs to be some other changes to make the visually hidden text work correctly (and indicate this to the user).

36degrees replied:
We could potentially leave this heading out entirely – it does follow naturally from the previous section. But makes it hard to deep link to it, for example.

calvin-lau-sig7 replied:
Shall we remove the heading @EoinShaughnessy? 
I do agree it follows naturally and the heading might be cause more confusion than it's worth. 

Also think the overall H3 seems to still work as-is.

EoinShaughnessy replied:
Yeah, reckon we should nix the heading. 👍🏻

EoinShaughnessy commented on src/components/summary-list/index.md.njk:
```diff
@@ -30,24 +30,31 @@ Do not use it for tabular data or a simple list of information or tasks, like a
+### Adding actions to only some rows
 
-{{ example({group: "components", item: "summary-list", example: "without-actions", html: true, nunjucks: true, open: false}) }}
+If you have a mix of rows with and without actions, add the `govuk-summary-list__row--no-actions` modifier class to the rows that do not have actions.
```
```suggestion
If you have a mix of rows with and without actions, add the `govuk-summary-list__row--no-actions` modifier class to the rows without actions.
```

calvin-lau-sig7 commented:
Okay, guidance looks to me thanks all! This is probably ready to move to 'Blocked' pending release of 4.0.

36degrees commented:
Rebased on top of the other guidance changes which I've split out into #1997 – that PR should be merged first at which point the base for this PR will update to `main`.

NikhilNanjappa commented:
Hi guys, is this not available on `govuk-frontend` v4.1.0 ?

Im using a "mixed actions" summary list macro and it doesnt seem to be adding the `govuk-summary-list__row--no-actions` class. Manually adding it using the `classes` also does not seem to remove the accessibility issue.

```
    {{ govukSummaryList({
      classes: 'govuk-!-margin-bottom-9',
      rows: [
        {
          key: { text: 'Key' },
          value: { text: 'Value' }
        }
      ]
    }) }}
```

36degrees commented:
@NikhilNanjappa this change was made in v4.0.0, so should work just fine in v4.1.0.

Can you confirm what HTML is being outputted by the macro?

NikhilNanjappa commented:
Hi @36degrees - this is the rendered HTML

```
<dl class="govuk-summary-list govuk-!-margin-bottom-9">
    ...
    <div class="govuk-summary-list__row my-class">
        <dt class="govuk-summary-list__key">
          Text
        </dt>
        <dd class="govuk-summary-list__value">
          Value
        </dd>
        <span class="govuk-summary-list__actions"></span>
    </div>
</dl>
```

36degrees commented:
Can I just check that you're definitely using the Nunjucks macros included in the GOV.UK Frontend package? You haven't got another version of the `govukSummaryList` macro in your codebase somewhere?

NikhilNanjappa commented:
We are using [GOV.UK CASA framework](https://github.com/dwp/govuk-casa/tree/7.0.7) and importing summary list as so:

```
{% from "govuk/components/summary-list/macro.njk" import govukSummaryList %}
```

36degrees commented:
Ahah, that would do it! 🙂 It looks like the latest version of GOV.UK CASA might support the latest version of GOV.UK Frontend, but you would need to upgrade to the latest version of GOV.UK CASA in your application.

If you're currently importing the CSS and JavaScript separately to GOV.UK CASA, make sure that you're using the same version of GOV.UK Frontend for the CSS, JavaScript and the Nunjucks macros and it should still display correctly.

Before v4.0.0 the summary list outputted an empty span (`<span class="govuk-summary-list__actions"></span>`) instead. Although this is technically invalid HTML, we are not aware of any real world issues that this would create.

NikhilNanjappa commented:
Unfortunately, I dont think we can upgrade just the govuk-frontend version without upgrading the CASA (which is a much bigger change and probably wouldnt be doing it)
